### PR TITLE
Use authentication for almost all endpoints

### LINKED ISSUE #162: Require JWT authentication for all routes
Except for some public routes like `login`, all routes should require a valid JSON web token for an existing user.

### PR DESCRIPTION
**Pull request by raffomania**
Fixes #162.

Things I've tested manually:

- Create a project and upload text + images
- Annotate a text file
- Create a memo for an annotation
- log in / log out
- run analysis
- create whiteboard with text document, code, memo
- update logbook
- read autologbook

I'm working on a small integration test that ensures we can't actually use the endpoints without a token, will update this PR once it's ready.

### CODE REVIEW

File: frontend/src/api/openapi/services/AnalysisService.ts [hunk 1]
```diff
@@ -122,7 +122,7 @@ export class AnalysisService {
       url: "/analysis/timeline_analysis",
       query: {
         project_id: projectId,
-        threshold: threshold / 100.0,
```
Comment (raffomania) [inline frontend/src/api/openapi/services/AnalysisService.ts:125]: Not sure where this change is coming from, @bigabig any idea?

File: backend/src/api/endpoints/user.py [hunk 2]
```diff
@@ -22,6 +27,8 @@
 from fastapi.security import OAuth2PasswordRequestForm
 from sqlalchemy.orm import Session
 
+# no authentication dependency for the whole group here,
+# as login and register need to be accessible withaut a token
```
Comment (raffomania) [inline backend/src/api/endpoints/user.py:26]: I'm thinking about moving the login and register route into a separate endpoint file, e.g. `auth.py` or similar. This way we could use the APIRouter in `user.py` to add the token dependency, like in the other endpoint files.

Comment (floschne) [reply backend/src/api/endpoints/user.py:26]: great idea! please do it

**Review by floschne — CHANGES_REQUESTED** [review]
Nice PR! Great that it works by just adding the `get_current_user` dependency to a router instead of all routes!

Please move on with your idea with the separate `auth` endpoint and inject `get_current_user` in the `user` endpoint.

File: backend/src/test/api/endpoints/test_general.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import pytest
+from httpx import AsyncClient
```
Comment (raffomania) [inline backend/src/test/api/endpoints/test_general.py:2]: We probably need to add this to `backend/environment.yml`?

File: backend/src/test/api/endpoints/test_authentication.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+from typing import Sequence, Set, Tuple
+
+import starlette.routing
+from fastapi.routing import APIRoute
+from main import app
+
+
+def test_authentication_required():
+    public_routes: Sequence[Tuple[Set[str], str]] = [
+        ({"GET"}, "/"),
```
Comment (floschne) [inline backend/src/test/api/endpoints/test_authentication.py:11]: please add `/heartbeat`

**Review by floschne — CHANGES_REQUESTED** [review]

**Review by floschne — COMMENTED** [review]
Nice PR! Did you test the app in the frontend?

File: backend/src/test/api/endpoints/test_general.py (hunk 3 above)
Comment (raffomania) [reply backend/src/test/api/endpoints/test_general.py:2]: When do I build the container, after or before this PR is merged?

**Comment by raffomania** [comment]
I think the secret detected by GitGuardinan is the JWT secret default value in `default.yaml`. Since we're probably overriding that in production (?), it should be fine.

**Comment by floschne** [comment]
> I think the secret detected by GitGuardinan is the JWT secret default value in `default.yaml`. Since we're probably overriding that in production (?), it should be fine.

Yes, it's fine because it's just the default and we generate a new one when deploying :)

File: frontend/src/api/openapi/services/AnalysisService.ts (hunk 1 above)
Comment (raffomania) [reply frontend/src/api/openapi/services/AnalysisService.ts:125]: I think we'll need to update the frontend code to perform the division by 100 elsewhere. The frontend is sending values between 0 and 100 right now.

**Comment by raffomania** [comment]
Tested again:


    Create a project and upload text + images
    Annotate a text file
    Create a memo for an annotation
    log in / log out
    run analysis
    create whiteboard with text document, code, memo
    update logbook
    read autologbook

Everything works fine :)

File: backend/src/test/api/endpoints/test_general.py (hunk 3 above)
Comment (floschne) [reply backend/src/test/api/endpoints/test_general.py:2]: It's better to build the container before just to make sure it builds and has no dep conflicts etc

Comment (raffomania) [reply backend/src/test/api/endpoints/test_general.py:2]: Oh, and I probably don't have access to the docker hub for pushing the image.

Comment (raffomania) [reply backend/src/test/api/endpoints/test_general.py:2]: Tried to build the container, got an error :D I'll investigate

Comment (raffomania) [reply backend/src/test/api/endpoints/test_general.py:2]: I didn't manage to build the container, but building it fails on the `mwp_v1` branch as well. So probably something is weird with my docker setup. 

The new packages in this PR are used for testing only, so maybe we can merge without pushing a new docker image?

Comment (raffomania) [reply backend/src/test/api/endpoints/test_general.py:2]: Managed to fix my setup :) The build now runs fine! I still don't think I can push to the registry due to missing access, though.

**Comment by floschne** [comment]
I also tested it a bit, and it seems to work nicely :) The only issue I have is that when I register a new user, log in, upload some docs, and want to annotate a doc, I get a "Generic Error" in the frontend, which comes from an HTTP 409 (Conflict) from the `PUT /adoc` route. If I remember correctly this was not the case before. If I understand it correctly, this happens because the route is called more than once with the same parameters and there can only be one adoc per user per document. 
Can you try to fix this please :)?

**Comment by raffomania** [comment]
> The only issue I have is that when I register a new user, log in, upload some docs, and want to annotate a doc, I get a "Generic Error" in the frontend, which comes from an HTTP 409 (Conflict) from the `PUT /adoc` route.

This one was a bit tricky! I managed to find the root cause which is an effect getting called multiple times during a single render. I managed to reproduce the issue on the `mwp_v1` branch as well. It's a super weird edge-case because having the effect fire twice in a single render means that all state is essentially unchanged and we can't use the usual reactive mechanisms to track if the request was already fired. I settled for adding a `useRef` which allows us to change values inside of a single render.

File: backend/src/test/api/endpoints/test_general.py (hunk 3 above)
Comment (floschne) [reply backend/src/test/api/endpoints/test_general.py:2]: Yes, sorry forgot that, I pushed it a few days ago :)

**Comment by floschne** [comment]
> > The only issue I have is that when I register a new user, log in, upload some docs, and want to annotate a doc, I get a "Generic Error" in the frontend, which comes from an HTTP 409 (Conflict) from the `PUT /adoc` route.
> 
> This one was a bit tricky! I managed to find the root cause which is an effect getting called multiple times during a single render. I managed to reproduce the issue on the `mwp_v1` branch as well. It's a super weird edge-case because having the effect fire twice in a single render means that all state is essentially unchanged and we can't use the usual reactive mechanisms to track if the request was already fired. I settled for adding a `useRef` which allows us to change values inside of a single render. It's now fixed in this PR :)

oh no this sounds indeed very complicated.. lucky us that you are another react magician ;)

**Pull request merged**